### ISSUE TITLE
chomp($data=<tied_fh>) strangeness.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5718]
Migrated from [rt.perl.org#10048](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=10048) (status was 'resolved')

Searchable as RT10048$

**Comment by p5pRT** [comment]
### From vkonovalov@peterstar.ru

#### Created by Administrator@localhost



Following program&#8203;:

package Foo;
sub TIEHANDLE \{ bless \{\}\, $\_\[0\]\}
sub READLINE \{return "foobar\\n"\}

package main;
\#use Foo;
tie\(\*TEST\, Foo\);
chomp\($data = \<TEST>\);
print join\("\, "\, map \{ ord\($\_\) \} split\(//\, $data\)\) \. "\\n";
chomp\($data\);
print join\("\, "\, map \{ ord\($\_\) \} split\(//\, $data\)\) \. "\\n";


results in strange output&#8203;:

102\, 111\, 111\, 98\, 97\, 114\, 10
102\, 111\, 111\, 98\, 97\, 114

but if instead of
chomp\($data = \<TEST>\);

write
chomp\($data = \<TEST>\, $data\);

or even
chomp\($data ||= \<TEST>\);

then program produces expected output&#8203;:

102\, 111\, 111\, 98\, 97\, 114
102\, 111\, 111\, 98\, 97\, 114

Affects both 5\.6\.1 and 5\.8\.0\-tobe

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by Administrator at Wed May  2 01:31:01 2001.

Summary of my perl5 (revision 5 version 6 subversion 1) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cl', ccflags
='-nologo -O1 -MD -DNDEBUG -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT 
 -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DPERL_MSVCRT_READFIX',
    optimize='-O1 -MD -DNDEBUG',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=4
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='', ldflags
'-nologo -nodefaultlib -release  -libpath:"d:\perl561\lib\CORE"  -machine:x8
6'
    libpth="D:\MSVStudio\VC98\LIB" "D:\MSVStudio\VC98\MFC\LIB"
"D:\apps\SQLLIB\LIB" "d:\perl561\lib\CORE"
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib
msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib
msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl56.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ',
ddlflags='-dll -nologo -nodefaultlib -release  -libpath:"d:\perl561\lib\CORE
"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.1:
    D:\Work\PerlScripts\utl
    D:\Work\PerlScripts\translations
    D:\Work\PerlScripts\sgml
    D:\Work\PerlScripts\pleps
    d:/perl561/lib
    d:/perl561/site/lib
    .


Environment for perl v5.6.1:
    HOME=d:\apps
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

PATH=d:\sh;d:\perl561\bin;d:\apps\vim;d:\apps\vslick-40\win;d:\bin;d:\cygwin
\bin;C:\WINNT\system32;C:\WINNT;C:\WINNT\System32\Wbem;d:\borland\cbuilder5\
bin
    PERL5CFG=d:\apps

PERL5LIB=D:\Work\PerlScripts\utl;D:\Work\PerlScripts\translations;D:\Work\Pe
rlScripts\sgml;D:\Work\PerlScripts\pleps
    PERL_BADLANG (unset)
    SHELL (unset)




```
</details>

**Comment by p5pRT** [comment]
### From @vanstyn

"Vadim Konovalov" \<vkonovalov@&#8203;peterstar\.ru> wrote&#8203;:
\[in synopsis\]
C\< chomp\($data = \<TEST>\); > does not chomp what is in $data if TEST
is a tied handle\.

This occurs because the handling of the tied case in do\_readline\(\)
does not pop the tied result and push back the target\, so the SV
we end up chomping is the temp returned from the tied READLINE
rather than the copy in $data\.

Attached patch fixes and tests it\. I suspect it isn't important enough
to force its way into 5\.8\, though\.

pp\_sys\.c has similar code for GETC handling\, but I was unable to
construct a failing test case for it so I haven't patched it\. In
principle an identical change should be required\.
No other lines matched /Set\.\*SV\.\*TOP/\.

Hugo
<details><summary>Inline Patch</summary>

```diff
--- pp_hot.c.old	Wed Jun 26 20:32:33 2002
+++ pp_hot.c	Sat Jul 13 17:18:20 2002
@@ -1454,8 +1454,11 @@
 	call_method("READLINE", gimme);
 	LEAVE;
 	SPAGAIN;
-	if (gimme == G_SCALAR)
-	    SvSetMagicSV_nosteal(TARG, TOPs);
+	if (gimme == G_SCALAR) {
+	    SV* result = POPs;
+	    SvSetSV_nosteal(TARG, result);
+	    PUSHTARG;
+	}
 	RETURN;
     }
     fp = Nullfp;
--- t/op/tiehandle.t.old	Sun Feb 17 12:43:04 2002
+++ t/op/tiehandle.t	Sat Jul 13 17:35:43 2002
@@ -77,7 +77,7 @@
 
 use Symbol;
 
-print "1..39\n";
+print "1..40\n";
 
 my $fh = gensym;
 
@@ -230,3 +230,15 @@
     Implement::compare(PRINT => @received);
 }
 
+{
+    # [ID 20020713.001] chomp($data=<tied_fh>)
+    local *TEST;
+    tie *TEST, 'CHOMP';
+    my $data;
+    chomp($data = <TEST>);
+    ok($data eq 'foobar');
+
+    package CHOMP;
+    sub TIEHANDLE { bless {}, $_[0] }
+    sub READLINE { "foobar\n" }
+}
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

> C\< chomp\($data = \<TEST>\); > does not chomp what is in $data if TEST
> is a tied handle\.
>
> This occurs because the handling of the tied case in do\_readline\(\)
> does not pop the tied result and push back the target\, so the SV
> we end up chomping is the temp returned from the tied READLINE
> rather than the copy in $data\.
>
> Attached patch fixes and tests it\. I suspect it isn't important enough
> to force its way into 5\.8\, though\.
>
> pp\_sys\.c has similar code for GETC handling\, but I was unable to
> construct a failing test case for it so I haven't patched it\. In
> principle an identical change should be required\.
> No other lines matched /Set\.\*SV\.\*TOP/\.

Wow\! thank you a lot for clarifying and even fixing that thing\! I will stop
my weak attempts to fix this\, and will learn from wise men\!

Vadim\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Hugo van der Sanden wrote&#8203;:
\[snip\]
> \+\{
> \+    \# \[ID 20020713\.001\] chomp\($data=\<tied\_fh>\)
> \+    local \*TEST;
> \+    tie \*TEST\, 'CHOMP';
> \+    my $data;
> \+    chomp\($data = \<TEST>\);

Due to how the interaction between tied\-ness and local\(\) isn't set in
stone\, I think that this testcase should have an explicit untie\(\)\.

> \+    ok\($data eq 'foobar'\);
> \+
> \+    package CHOMP;
> \+    sub TIEHANDLE \{ bless \{\}\, $\_\[0\] \}
> \+    sub READLINE \{ "foobar\\n" \}
> \+\}

\-\- 
tr/\`4/ /d\, print "@&#8203;\{\[map \-\-$| ? ucfirst lc : lc\, split\]\}\,\\n" for
pack 'u'\, pack 'H\*'\, 'ab5cf4021bafd28972030972b00a218eb9720000';

**Comment by p5pRT** [comment]
### From @hvds

Earlier I wrote&#8203;:
:"Vadim Konovalov" \<vkonovalov@&#8203;peterstar\.ru> wrote&#8203;:
:\[in synopsis\]
:C\< chomp\($data = \<TEST>\); > does not chomp what is in $data if TEST
:is a tied handle\.
:
:This occurs because the handling of the tied case in do\_readline\(\)
:does not pop the tied result and push back the target\, so the SV
:we end up chomping is the temp returned from the tied READLINE
:rather than the copy in $data\.

This applied as \#17716\, and should also fix \[perl \#15886\]\.

Hugo

**Comment by p5pRT** [comment]
@gbarr - Status changed from 'open' to 'resolved'
